### DISCUSSION TITLE
Building OpenRoad locally on an Intel mac

### DISCUSSION THREAD

**Discussion by ishwars1618** [The-OpenROAD-Project/OpenROAD-flow-scripts#637] [General]
Hello, I'm a student learning digital design and would like to use OpenRoad.  I am trying to build OpenRoad locally on my Macbook Pro (Catalina MacOS 10.15). I have installed pandas, tcl, libffi etc. using brew, and did a git clone to get the build scripts. However, when building it, I run into first "readlink: illegal option -- f", and then errors below (after which I aborted the build).

kernel/driver.cc:178:4: error: use of undeclared identifier 'append_history'
                        append_history(where_history() - yosys_history_offset, yosys_history_file.c_str());
                        ^
kernel/driver.cc:188:27: error: use of undeclared identifier 'history_list'
        HIST_ENTRY **hist_list = history_list();

Not sure how to fix the errors.
Can anyone share a complete setup for building OpenRoad locally on a Mac?
Thanks.

**Comment by maliberty** [comment]
@vvbandeira @rovinski

**Comment by rovinski** [comment]
* Have you run the dependency installer script at `OpenROAD-flow-scripts/tools/OpenROAD/etc/DependencyInstaller.sh`?
* Did you update your `PATH` using the suggested commands from the installer?
* Can you please provide your complete build log? The errors look like they're coming from yosys, not openroad.

**Reply by ishwars1618** [reply]
Thanks for the suggestions. I ran the dependency installer script and then updated the PATH.  Before that I also performed updates on a few utils using brew.  The error is a bit different now (ffi related).  

frontends/ast/dpicall.cc:25:10: fatal error: 'ffi.h' file not found
#include <ffi.h>
         ^~~~~~~
1 error generated.
make: *** [frontends/ast/dpicall.o] Error 1

Also attached the log.  [](url)
[build.log](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/files/9956461/build.log)

**Reply by rovinski** [reply]
There are a few packages for yosys that are unfortunately not installed by the dependency installer (the installer is only for OpenROAD).

You should be able to get the remaining packages with:
```
brew install libffi pkg-config klayout
```
I think MacOS marks klayout as an untrusted app and you'll need to mark it as trusted before it will run.

**Reply by ishwars1618** [reply]
Thanks for the suggestion! I installed those with brew, and the rebuild went through almost the entire build process until it reached this error below.  Attached the CMakeError.log as well.
[CMakeError.log](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/files/9965457/CMakeError.log)

CMake Error at /usr/local/Cellar/cmake/3.24.2/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES)
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.24.2/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  /usr/local/Cellar/cmake/3.24.2/share/cmake/Modules/FindOpenMP.cmake:545 (find_package_handle_standard_args)
  src/drt/CMakeLists.txt:48 (find_package)

-- Configuring incomplete, errors occurred!
See also "/Users/ishwork/Documents/OR/OpenROAD-flow-scripts/tools/OpenROAD/build/CMakeFiles/CMakeOutput.log".
See also "/Users/ishwork/Documents/OR/OpenROAD-flow-scripts/tools/OpenROAD/build/CMakeFiles/CMakeError.log".

**Reply by rovinski** [reply]
It is more helpful to attach the CMake log rather than just just CMakeError.log.

Do you have Apple Clang installed? I don't think we support GCC on mac (at least, it is not well test). Otherwise the error means that you need to use a compiler that has OpenMP installed.

**Reply by ishwars1618** [reply]
I installed the Xcode command-line tools successfully - so assuming Apple Clang comes with it. Attaching the CMakeOutput.log now.  What do I need to use a compiler that has OpenMP installed?

[CMakeOutput.log](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/files/9967453/CMakeOutput.log)

**Reply by ishwars1618** [reply]
Verified that Clang is installed (please see below)
% clang --version
Apple clang version 12.0.0 (clang-1200.0.32.29)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin

**Reply by vvbandeira** [reply]
@ishwars1618 as Austin mentioned, the DependencyInstaller.sh is a good option to automatically install the dependencies, you can check a more manual approach [here](https://github.com/The-OpenROAD-Project/OpenROAD/blob/ff442dd5425fdd179418de62187f52620d0939a9/.github/workflows/github-actions-macos.yml#L19-L45). Note the exports and how to call `cmake`.
